### ISSUE TITLE
Sandbox.reset is not resetting call count for Stubs

### ISSUE THREAD

**Issue by madSimonJ** [sinonjs/sinon#1089]
* Sinon version : ^1.17.4
* Environment   : NodeJS 6.2.1
* Example URL   : https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/DataAccess/repositories/bookRespository.spec.js
* Other libraries you are using: chai, sinon-chai,chai-as-promised, mockery

**What did you expect to happen?**

To create a Sandbox with sinon.sandbox.create(), and then calling its stub() function to create a stub, and having it added to an internal collection, ,which can have all contained call counts reset at once by calling sandbox.reset()

**What actually happens**

No errors occur, but the call counts do not reset.

**How to reproduce**

The following is an example of code I've written with sinon:

https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/DataAccess/repositories/bookRespository.spec.js

The creation of the Sinon Stubs is in this module:

https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/stubs/mongoDBStub.js

The code to call the reset method is here:

https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/stubs/TestHelper.js

In bookRepository.spec.js, the unit test here:

https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/DataAccess/repositories/bookRespository.spec.js#L99

fails unless I include an additional line in the After function of the previous set of tests to specifically reset the individual Stub I'm asserting against:

https://github.com/madSimonJ/protractor-demo-gtx/blob/master/specs/DataAccess/repositories/bookRespository.spec.js#L48

Without that line, the test fails, saying the Stub was called, when it should not have been in that test pass.

**Comment by jonnyreeves** [comment]
I just tried to create a minimal test-case to reproduce this issue but was unsuccessful; please can you isolate the issue to a smaller code-sample (or preferably a new sinon test-case).

FYI, here's the code I used:

```js
var sb = sinon.sandbox.create();
var s = sb.stub();
s();
s();
console.log(s.callCount); // 2
sb.reset();
console.log(s.callCount); // 0
```

**Comment by fatso83** [comment]
I have also looked into this quite extensively, and I cannot see anything _immediately_ wrong with your code. I actually cloned the project and ran the tests to see if I could understand it better, but alas, it seems fine. It is hard for us to justify spending more time on digging into this from our side, as our basic tests seem fine, but if I were you I'd try to delete code block after code block to the point you are stuck with the minimal set of lines needed to reproduce it. 

P.S. You might consider adding a README and a `"test : gulp test"` in your `package.json` to make it faster for others to get it running ;-)

**Comment by fatso83** [comment]
OK, I actually could not keep away. This is basically user error, I just did not see what. 

See this [little snippet](https://tonicdev.com/5783c1e8ed16711300a08d62/5783c4440717d21200ab6b9e) I made for debugging.

You run `verifyAndRestore()` on the sandbox, which will remove all the stubs from its internal collection. You then try to reset the same sandbox (which is now empty), to no effect.

Try just using `verify`.
